### PR TITLE
Add MessageActions/SessionMessageActions

### LINKED ISSUE #18361: [WebJobs/ServiceBus] Remove use of reflection to bind to Receiver
I don't think we should get too focused on the reflection perf here. This needs to be fixed to use a public API.

_Originally posted by @pakrym in https://github.com/Azure/azure-sdk-for-net/pull/18330#discussion_r568761819_

### LINKED ISSUE #18408: [WebJobs/ServiceBus] Update binding parameter names
Add bindings for the properties that differ from Track 1, e.g. ExpiresAtUtc, EnqueuedTmeUtc, and Subject. Also add tests to make sure binding works for these and the old names.

See if it is possible to update the binding names for "MessageReceiver" and "MessageSession" as well.

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Fixes https://github.com/Azure/azure-sdk-for-net/issues/18408 and https://github.com/Azure/azure-sdk-for-net/issues/18361

### CODE REVIEW

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/api/Microsoft.Azure.WebJobs.Extensions.ServiceBus.netstandard2.0.cs [hunk 1]
```diff
@@ -65,6 +65,15 @@ public partial class MessagingProvider
         public virtual Azure.Messaging.ServiceBus.ServiceBusProcessor CreateProcessor(string entityPath, string connectionString) { throw null; }
         public virtual Microsoft.Azure.WebJobs.ServiceBus.SessionMessageProcessor CreateSessionMessageProcessor(string entityPath, string connectionString) { throw null; }
     }
+    public partial class ServiceBusMessageActions
+    {
+        internal ServiceBusMessageActions() { }
```
Comment (JoshLove-msft) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/api/Microsoft.Azure.WebJobs.Extensions.ServiceBus.netstandard2.0.cs:70]: Should add protected ctor for mocking (or make ctors public).

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusHostBuilderExtensions.cs [hunk 2]
```diff
@@ -75,6 +75,8 @@ public static IWebJobsBuilder AddServiceBus(this IWebJobsBuilder builder, Action
                         "SessionHandlerOptions:MaxConcurrentSessions",
                         options.MaxConcurrentSessions);
 
+                    options.SessionIdleTimeout = section.GetValue("SessionHandlerOptions:MessageWaitTime", options.SessionIdleTimeout);
```
Comment (JoshLove-msft) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusHostBuilderExtensions.cs:78]: This is now supported in the SDK - SessionIdleTimeout.

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusOptions.cs [hunk 3]
```diff
@@ -131,19 +131,20 @@ public int MaxConcurrentSessions
         /// </summary>
         public Func<ProcessErrorEventArgs, Task> ExceptionHandler { get; set; }
 
-        /// <summary>
-        /// Gets or sets an optional <see cref="TimeSpan"/> specifying the maximum time to wait when attempting to receive messages.
-        /// If not specified, the <see cref="ServiceBusRetryOptions.TryTimeout"/> will be used. This only applies for functions that receive
-        /// a batch of messages.
-        /// </summary>
-        public TimeSpan? MaxWaitTime { get; set; }  // TODO this should probably be exposed in the Processor as well and would then also apply for functions that receive a single message.
```
Comment (JoshLove-msft) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusOptions.cs:139]: This was never actually exposed in the T1 SDK/extensions for non-sessions.

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/api/Microsoft.Azure.WebJobs.Extensions.ServiceBus.netstandard2.0.cs (hunk 1 above)
Comment (pakrym) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/api/Microsoft.Azure.WebJobs.Extensions.ServiceBus.netstandard2.0.cs:70]: Make it abstract?

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Triggers/ServiceBusTriggerBindingStrategy.cs [hunk 4]
```diff
@@ -53,8 +52,12 @@ public ServiceBusReceivedMessage[] BindMultiple(ServiceBusTriggerInput value, Va
             }
 
             var bindingData = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
-            SafeAddValue(() => bindingData.Add("MessageReceiver", value.Receiver));
-            SafeAddValue(() => bindingData.Add("MessageSession", value.SessionReceiver));
+            // TODO - investigate why the parameter names need to be hard-coded here but they are not
+            // for binding to sender
+            SafeAddValue(() => bindingData.Add("MessageReceiver", value.MessageActions));
```
Comment (pakrym) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Triggers/ServiceBusTriggerBindingStrategy.cs:57]: Is it worth supporting the old names if we change the type?

Comment (JoshLove-msft) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Triggers/ServiceBusTriggerBindingStrategy.cs:57]: I thought we wanted to do this.. but I don't think so.

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusExtensionConfigProvider.cs [hunk 5]
```diff
@@ -89,6 +89,9 @@ public void Initialize(ExtensionConfigContext context)
             // register our binding provider
             ServiceBusAttributeBindingProvider bindingProvider = new ServiceBusAttributeBindingProvider(_nameResolver, _options, _configuration, _messagingProvider);
             context.AddBindingRule<ServiceBusAttribute>().Bind(bindingProvider);
+
+            context.AddBindingRule<ServiceBusAttribute>()
```
Comment (JoshLove-msft) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusExtensionConfigProvider.cs:93]: This should be removed.

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Primitives/MessageProcessor.cs [hunk 6]
```diff
@@ -36,11 +36,11 @@ public MessageProcessor(ServiceBusProcessor processor)
         /// This method is called when there is a new message to process, before the job function is invoked.
         /// This allows any preprocessing to take place on the message before processing begins.
         /// </summary>
-        /// <param name="receiver"></param>
+        /// <param name="messageActions"></param>
```
Comment (jsquire) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Primitives/MessageProcessor.cs:39]: Empty doc comments make me sad.

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Primitives/ServiceBusMessageActions.cs [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs.ServiceBus
+{
+    /// <summary>
+    /// Represents the set of message actions that can be performed on a <see cref="ServiceBusReceivedMessage"/>.
+    /// </summary>
+    public class ServiceBusMessageActions
```
Comment (jsquire) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Primitives/ServiceBusMessageActions.cs:14]: This is a clever approach to normalizing.   I like it.

**Review by jsquire — APPROVED** [review]
LGTM, but I'm not confident that I have a full understanding of the context;  I'd trust @pakrym's thoughts over mine here.

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Primitives/ServiceBusMessageActions.cs (hunk 7 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Primitives/ServiceBusMessageActions.cs:14]: Yeah I wonder who could have come up with it 😆

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusExtensionConfigProvider.cs (hunk 5 above)
Comment (pakrym) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusExtensionConfigProvider.cs:93]: ??

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Triggers/ServiceBusTriggerBindingStrategy.cs (hunk 4 above)
Comment (pakrym) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Triggers/ServiceBusTriggerBindingStrategy.cs:57]: I think it makes things more confusing and doesn't allow us to add the MessageReceiver type back if required.

**Review by pakrym — APPROVED** [review]

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Primitives/ServiceBusMessageActions.cs (hunk 7 above)
Comment (pakrym) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Primitives/ServiceBusMessageActions.cs:14]: Should we add an inheritance relationship between ServiceBusMessageActions and SessionActions?

Comment (pakrym) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Primitives/ServiceBusMessageActions.cs:14]: We already have a bunch of if-elses we would need to add another one for session.

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusExtensionConfigProvider.cs (hunk 5 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusExtensionConfigProvider.cs:93]: I just haven't gotten around to it yet..

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Triggers/ServiceBusTriggerBindingStrategy.cs (hunk 4 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Triggers/ServiceBusTriggerBindingStrategy.cs:57]: I'm not sure I agree. I can see people going through working functions and just swapping out the types. I think we will have to decide on the Receiver before GA - I don't think it is something we would add post-GA. One other thing I need to investigate is why the binding requires a specific parameter name to begin with. For the sender, you can use any parameter name.

Comment (pakrym) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Triggers/ServiceBusTriggerBindingStrategy.cs:57]: Is it only the type? Do all method names and parameters match?

Comment (pakrym) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Triggers/ServiceBusTriggerBindingStrategy.cs:57]: I mean with track 1. If customers would need to change the parameter name and usages changing the name is not a problem.

Comment (JoshLove-msft) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Triggers/ServiceBusTriggerBindingStrategy.cs:57]: That is what I am proposing - that they wouldn't need to change the parameter name. Which name are you referring to? The Type name?

**Pull request merged**